glutin/src/api/egl/display.rs

```diff
@@ -499,6 +502,19 @@ impl Display {
         });
         Ok(Self { inner })
     }
+
+    fn sanitize_libglvnd_aliasing(display: EglDisplay, version: Version) -> EglDisplay {
+        match (display, version) {
+            // libglvnd has an unsavoury quirk of aliasing `getPlatformDisplay` and
+            // `getPlatformDisplayEXT`, which may result in an `EglDisplay::Khr` being created
+            // for a platform that does not support KHR functions and only supports EXT
+            // functions. As such, we check if the version is equal to 1.4 and downgrade to
+            // `EglDisplay::Ext` if so.
+            (EglDisplay::Khr(display), Version { major: 1, minor: 4 }) => EglDisplay::Ext(display),
```

Comment:
I was thinking to ensure that the function pointers for `egl.GetPlatformDisplay` and `egl.GetPlatformDisplayExt` are the same before doing such thing, so you can ensure that the cast is _safe_ because the same functions were used to produce the pointer.

Just doing a raw cast might be not a great idea if the functions are _different_?

Reply:
> I was thinking to ensure that the function pointers for `egl.GetPlatformDisplay` and `egl.GetPlatformDisplayExt` are the same before doing such thing, so you can ensure that the cast is _safe_ because the same functions were used to produce the pointer.

The problem with that is that they are only aliased by `libglvnd` internally. The functions requested through the dynamic library are technically distinct, but otherwise call the same underlying vendor function.

> Just doing a raw cast might be not a great idea if the functions are different?

I'd say that considering the version is explicitly checked here to be 1.4 before the downgrade, this isn't much of a problem per `libglvnd`'s aliasing and the fact that `EglDisplay::Khr` shouldn't even be possible for 1.4. Although, I suppose this could cause EGL to error out at surface creation if `EGL_EXT_platform_base` isn't actually available for the platform, but that hinges on the KHR display connection request not failing beforehand, and I'm not entirely sure how that would play out when the underlying vendor function is `eglGetDisplay` or whether that's even possible.

Reply:
>The problem with that is that they are only aliased by libglvnd internally. The functions requested through the dynamic library are technically distinct, but otherwise call the same underlying vendor function.

We can not assume that, if pointers are different then we not do a casting since they could in fact be different. The problem is that libglvnd tells us that 1.5 exists due to one vendor providing it, while it's not due to issue in how it's done in libglvnd. When using only a vendor library you won't ever have this issue in the first place from what I understand.

So, ensure the pointers are the same and that's the only time we can do such a cast safely.

Reply:
> We can not assume that, if pointers are different then we not do a casting since they could in fact be different.

`libglvnd`'s public functions, the functions that are actually used through the dynamic library, are distinct (as the aliasing is only done internally in `libglvnd`; the public functions call the same underlying vendor function), so the check would inevitably fail and the downgrade wouldn't happen.

> When using only a vendor library you won't ever have this issue in the first place from what I understand.

Likely not, as vendor libraries would, I presume, only have the functions that are relevant to them.

> The problem is that libglvnd tells us that 1.5 exists due to one vendor providing it

Well, not quite. Firstly, `libglvnd` advertises client extensions if at least one implementation supports them, so fully discriminating based on them is out of the question. Secondly, `eglGetPlatformDisplay` and `eglGetPlatformDisplayEXT` are functions that exist in `libglvnd` and are loaded regardless of whether any of the platforms support them, and both internally call the same underlying vendor function, so fully discriminating based on whether they are loaded is also out of the question for `libglvnd`. If the display creation manages to succeed, we are basically only left to do a check for the version after display initialization, which the PR already introduces.

Reply:
Hm, indeed they'll be different. I was thinking they just provide the same pointer when loading but no, they use different pointers but call to the same implementation.

In this case, I'd suggest to this only for `#[cfg(free_unix)]`.

I'll also open an issue upstream.

Reply:
Separate question: is EGL `1.3` or older something that even exists out in the wild and that `glutin` could possibly run on (via `libglvnd`)? If so, should we do this?


```suggestion
            (EglDisplay::Khr(display), Version { major, minor }) if major <= 1 && minor <= 4 => EglDisplay::Ext(display),
```

Reply:
> If so, should we do this?

Probably not, since [`EGL_EXT_platform_base`](https://registry.khronos.org/EGL/extensions/EXT/EGL_EXT_platform_base.txt), the extension that provides the functions for `Ext` displays, is said in the specification to require 1.4. Downgrading to `Legacy` if the version number is even lower than 1.4 could be done, but I'm not certain how 1.3 or older would behave in this situation.